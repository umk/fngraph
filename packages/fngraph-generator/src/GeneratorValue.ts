import { Context } from '@fngraph/data'

type GeneratorValue = {
  /**
   * An index of node in an ordered array of nodes that has
   * advanced the last time.
   */
  n: number
  /**
   * The contexts that has been accepted the last time by
   * respective nodes in an ordered collection of nodes. A node
   * can be undefined if the context declarations conflict with
   * declarations of its parent contexts.
   */
  contexts: GeneratorValueContexts
}

export type GeneratorValueContexts = Array<Context | undefined>

export default GeneratorValue
