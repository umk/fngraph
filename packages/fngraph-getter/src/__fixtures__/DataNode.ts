import { DataNode } from '@fngraph/generator'

export function createDataNode(overrides?: Partial<DataNode>): DataNode {
  return {
    incoming: [],
    outgoing: [],
    getter: jest.fn(),
    priority: 0,
    invert: false,
    getProperties: () => [],
    ...overrides,
  }
}
