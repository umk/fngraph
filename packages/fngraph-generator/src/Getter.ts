import { DataRecord } from '@fngraph/data'

import GeneratorContext from './GeneratorContext'
import GeneratorValue from './GeneratorValue'
import GetContextRecord from './GetContextRecord'
import IncomingMapper from './IncomingMapper'
import OutgoingMapper from './OutgoingMapper'
import PropertyRef from './PropertyRef'

/**
 * Creates a getter function with incoming and outgoing declarations
 * that respective mappers operate
 * @param incoming The incoming declarations mapper
 * @param outgoing The outgoing declarations mapper
 * @returns A getter function
 */
export type GetterFactory<P extends DataRecord, R extends DataRecord> = (
  incoming: IncomingMapper<P>,
  outgoing: OutgoingMapper<R>,
) => Getter

/**
 * A function, which iterates over the records and derives new
 * data against each of the record. Upon yield the getter must
 * update the indexes array in a context to reflect global unique
 * location of the resulting record.
 * @param properties A collection of properties to be collected from the getter output
 * @param records A collection of records over which to derive the values
 * @param context A node iterator context
 * @param getRecord A function which gets a context record from generator value
 * @param invert Indicates whether the getter output is inverted
 */
type Getter = (
  properties: Array<PropertyRef>,
  records: AsyncGenerator<GeneratorValue>,
  context: GeneratorContext,
  getRecord: GetContextRecord,
  invert: boolean,
) => AsyncGenerator<GeneratorValue>

/**
 * The default max number of items to be fed into a batched
 * getter.
 */
export const DEFAULT_MAX_BATCH = 10

/**
 * Default number of generator values preloaded into buffer
 * in order to extract unique records that fed into a batched
 * getter.
 */
export const DEFAULT_MAX_BUFFER = 1000

export default Getter
