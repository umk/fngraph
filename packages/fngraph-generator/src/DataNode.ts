import { DeclarationID } from '@fngraph/data'

import Getter from './Getter'

type DataNode = {
  /**
   * Declarations which values the node depends on.
   */
  incoming: Array<DeclarationID>
  /**
   * Declarations which values are produced by the node.
   */
  outgoing: Array<DeclarationID>
  /**
   * A function that actually produces declarations based on
   * combinations of incoming declarations.
   */
  getter: Getter
  /**
   * Indicates whether node output is inverted. The inverted
   * nodes cannot have outgoing declarations as they just
   * serve as a predicate for incoming declarations.
   */
  invert: boolean
  /**
   * The priority the node is invoked. Higher number means the
   * node is invoked later.
   */
  priority: number
}

export default DataNode
