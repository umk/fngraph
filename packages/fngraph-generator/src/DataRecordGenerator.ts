import { DataRecord, Declaration, DeclarationID } from '@fngraph/data'

import DataNode from './DataNode'
import DataNodeSequence from './DataNodeSequence'
import GeneratorContext from './GeneratorContext'
import GeneratorValue from './GeneratorValue'
import { createGetContextRecord } from './GetContextRecord'
import { createMergeContexts } from './MergeContexts'

type DataRecordGenerator = () => AsyncGenerator<DataRecord>

export function createDataRecordGenerator(
  sequenceOrNodes: DataNodeSequence | Array<DataNode>,
  declarations: Array<Declaration>,
): DataRecordGenerator {
  const sequence = Array.isArray(sequenceOrNodes)
    ? DataNodeSequence.create(sequenceOrNodes)
    : sequenceOrNodes
  const parentRefs = getParentRefs(sequence.nodes, sequence.incomings)
  const parentTrRefs = getParentTrRefs(sequence.nodes, parentRefs)
  const incomingDecls = Array.from(
    sequence.nodes.reduce((prev, cur) => {
      cur.incoming.forEach((d) => prev.add(d))
      return prev
    }, new Set<DeclarationID>()),
  )
  const mergeContexts = createMergeContexts(declarations)
  function createGeneratorAt(index: number): {
    iter: () => AsyncGenerator<GeneratorValue>
    isCached: boolean
  } {
    if (index === sequence.nodes.length) {
      let iter: () => AsyncGenerator<GeneratorValue>
      if (index === 0) {
        iter = async function* (): AsyncGenerator<GeneratorValue> {
          // Don't return anything
        }
      } else {
        iter = async function* (): AsyncGenerator<GeneratorValue> {
          yield {
            contexts: new Array(sequence.nodes.length).fill(undefined),
            n: sequence.nodes.length,
          }
        }
      }
      return { iter, isCached: false }
    }
    const upstreamIter = createGeneratorAt(index + 1)
    const currentRefs = parentRefs[index]
    const currentTrRefs = parentTrRefs[index].sort((a, b) => b - a)
    // Cache the node output if...
    const isCached =
      // ...upstream node is cached, or...
      upstreamIter.isCached ||
      // ...current node doesn't depend on anything, but is not the last node, or...
      (currentTrRefs.length === 0 && index < sequence.nodes.length - 1) ||
      // ...current node has a dependency, which is not an upstream node
      currentTrRefs[0] > index + 1
    const context = new GeneratorContext(index, currentTrRefs, isCached)
    const { getter, getProperties, invert } = sequence.nodes[index]
    const getRecord = createGetContextRecord(currentRefs, mergeContexts)
    const properties = getProperties(incomingDecls, invert)
    const iter = function (): AsyncGenerator<GeneratorValue> {
      const records = upstreamIter.iter()
      return getter(properties, records, context, getRecord, invert)
    }
    return { iter, isCached }
  }
  const outputRefs = getOutputRefs(sequence.nodes, parentRefs)
  const getRecord = createGetContextRecord(outputRefs, mergeContexts)
  return async function* (): AsyncGenerator<DataRecord> {
    const current = createGeneratorAt(0)
    for await (const value of current.iter()) {
      const record = getRecord(value.contexts)
      if (record) yield record
    }
  }
}

function getParentRefs(
  nodes: Array<DataNode>,
  incomings: Map<DeclarationID, Array<DataNode>>,
): Array<Array<number>> {
  const parents = nodes.reduce(
    (prev, cur, i) => {
      for (const outgoing of cur.outgoing) {
        const incoming = incomings.get(outgoing)
        if (incoming) {
          for (const node of incoming) {
            const n = nodes.indexOf(node)
            prev[n].add(i)
          }
        }
      }
      prev.push(new Set())
      return prev
    },
    [] as Array<Set<number>>,
  )
  return parents.map((p) => Array.from(p))
}

function getParentTrRefs(
  nodes: Array<DataNode>,
  parentRefs: Array<Array<number>>,
): Array<Array<number>> {
  const r = parentRefs.map((u) => new Set<number>(u))
  for (let i = nodes.length - 1; i >= 0; i--) {
    const current = r[i]
    parentRefs[i].forEach((u) => {
      r[u].forEach((n) => current.add(n))
    })
  }
  return r.map((p) => Array.from(p))
}

function getOutputRefs(nodes: Array<DataNode>, parentRefs: Array<Array<number>>) {
  const incomings = new Array(nodes.length).fill(0)
  parentRefs.forEach((r) => r.forEach((n) => incomings[n]++))
  return incomings.reduce((prev, cur, i) => {
    if (cur === 0) prev.push(i)
    return prev
  }, [] as Array<number>)
}

export default DataRecordGenerator
