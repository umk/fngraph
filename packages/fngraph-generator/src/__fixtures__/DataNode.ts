import DataNode from '../DataNode'

export function createDataNode(overrides?: Partial<DataNode>): DataNode {
  return {
    incoming: [],
    outgoing: [],
    getter: jest.fn(),
    priority: 0,
    invert: false,
    getProperties: () => [],
    ...overrides,
  }
}
