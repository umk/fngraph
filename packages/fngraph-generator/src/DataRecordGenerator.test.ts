import { DataRecord, Declaration, DeclarationID } from '@fngraph/data'

import DataNode from './DataNode'
import { createDataRecordGenerator } from './DataRecordGenerator'
import GeneratorValue from './GeneratorValue'
import Getter from './Getter'
import { createDataNode } from './__fixtures__'

function createGetter(values: Array<DataRecord>): Getter {
  return async function* (
    _properties,
    records,
    context,
    getRecord,
    invert,
  ): AsyncGenerator<GeneratorValue> {
    if (invert) throw new Error('inversion is not supported')
    for await (const current of records) {
      context.advanceToValueGroup(current)
      const base = getRecord(current.contexts)
      if (base) {
        let n = current.n
        for (const v of values) {
          current.contexts[context.nodeIndex] = { ...base, ...v }
          yield { contexts: current.contexts, n }
          n = context.nodeIndex
        }
      }
    }
  }
}

describe('createDataRecordGenerator', () => {
  const declarationA = 'declaration A' as DeclarationID
  const declarationB = 'declaration B' as DeclarationID
  const declarationC = 'declaration C' as DeclarationID

  it('throws an error if nodes have recursive dependencies', () => {
    const nodeA: DataNode = createDataNode({ incoming: [declarationA], outgoing: [declarationB] })
    const nodeB: DataNode = createDataNode({ incoming: [declarationB], outgoing: [declarationA] })

    expect(() => createDataRecordGenerator([nodeA, nodeB], [])).toThrowErrorMatchingSnapshot()
  })
  it('throws an error nodes has a dependency on itself', () => {
    const nodeA: DataNode = createDataNode({ incoming: [declarationA], outgoing: [declarationA] })

    expect(() => createDataRecordGenerator([nodeA], [])).toThrowErrorMatchingSnapshot()
  })
  it("doesn't return anything if there are no nodes", async () => {
    const nodes: Array<DataNode> = []
    const declarations: Array<Declaration> = []

    const generator = createDataRecordGenerator(nodes, declarations)
    const records = generator()

    const { done } = await records.next()

    expect(done).toBe(true)
  })
  it('iterates through getter records', async () => {
    const nodeA = createDataNode({
      outgoing: [declarationA, declarationB],
      getter: createGetter([
        { [declarationA]: 1, [declarationB]: 2 },
        { [declarationA]: 10, [declarationB]: 20 },
      ]),
    })
    const nodeB = createDataNode({
      incoming: [declarationA, declarationB],
      outgoing: [declarationC],
      getter: createGetter([
        { [declarationC]: 3 },
        { [declarationC]: 30 },
        { [declarationC]: 300 },
      ]),
    })
    const generator = createDataRecordGenerator([nodeA, nodeB], [])
    const records: Array<DataRecord> = []
    for await (const record of generator()) records.push(record)
    expect(records).toMatchSnapshot()
  })
})
