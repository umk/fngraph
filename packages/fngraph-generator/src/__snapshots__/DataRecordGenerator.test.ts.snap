// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createDataRecordGenerator iterates through getter records 1`] = `
[
  {
    "declaration A": 1,
    "declaration B": 2,
    "declaration C": 3,
  },
  {
    "declaration A": 1,
    "declaration B": 2,
    "declaration C": 30,
  },
  {
    "declaration A": 1,
    "declaration B": 2,
    "declaration C": 300,
  },
  {
    "declaration A": 10,
    "declaration B": 20,
    "declaration C": 3,
  },
  {
    "declaration A": 10,
    "declaration B": 20,
    "declaration C": 30,
  },
  {
    "declaration A": 10,
    "declaration B": 20,
    "declaration C": 300,
  },
]
`;

exports[`createDataRecordGenerator throws an error if nodes have recursive dependencies 1`] = `"The nodes have recursive dependency"`;

exports[`createDataRecordGenerator throws an error nodes has a dependency on itself 1`] = `"The nodes have recursive dependency"`;
